pyhap/accessory_driver.py

```diff
@@ -157,6 +181,68 @@ def __init__(self, *, address=None, port=51234,
         network_tuple = (self.state.address, self.state.port)
         self.http_server = HAPServer(network_tuple, self)
 
+    def start(self):
+        """Start the event loop and call `start_pyhap`.
+
+        Pyhap will be stopped gracefully on a KeyBoardInterrupt.
+        """
+        try:
+            logger.info('Starting the event loop')
+            self.add_job(self.start_pyhap)
+            self.loop.run_forever()
+        except KeyboardInterrupt:
+            self.loop.call_soon_threadsafe(
+                self.loop.create_task, self.async_stop())
+            self.loop.run_forever()
+        finally:
+            self.loop.close()
+            logger.info('Closed the event loop')
+
+    def stop(self):
+        """Method to stop pyhap."""
+        self.loop.call_soon_threadsafe(
+            self.loop.create_task, self.async_stop())
+
+    async def async_stop(self):
+        """Stops the AccessoryDriver and shutdown all remaining tasks."""
+        await self.async_add_job(self.stop_pyhap)
+        logger.debug('Shutdown executers')
+        self.executer.shutdown()
+        self.loop.stop()
+
+    def add_job(self, target, *args):
+        """Add job to executer pool."""
+        if target is None:
+            raise ValueError("Don't call add_job with None.")
+        self.loop.call_soon_threadsafe(self.async_add_job, target, *args)
+
+    @callback
+    def async_add_job(self, target, *args):
+        """Add job from within the eventloop."""
+        task = None
+
+        if asyncio.iscoroutine(target):
+            task = self.loop.create_task(target)
+        elif is_callback(target):
+            self.loop.call_soon(target, *args)
+        elif asyncio.iscoroutinefunction(target):
+            task = self.loop.create_task(target(*args))
+        else:
+            task = self.loop.run_in_executor(None, target, *args)
+
+        return task
+
+    @callback
+    def async_run_job(self, target, *args):
```

Comment:
Can you elaborate on this? The idea here is to be able to execute a `@callback` immediately right? Why not let the client just run its immediate method if she knows it is not async.

Reply:
In theory yes, however what if we want to change a method from `callback` to `async` or `normal`? We would need to change all calls as well. Calling these methods through the helpers, even for direct callback, simplifies that.

Reply:
Okay makes sense